File: ChartJs3.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using Newtonsoft.Json.Linq;
using System;
using System.Threading.Tasks;

namespace NC.Blazor
{
    public partial class ChartJs3 : IAsyncDisposable
    {
        public class Data
        {
            public object x { get; set; }

            public double y { get; set; }

            public Data(object label, double value)
            {
                this.x = label;
                this.y = value;
            }
        }

        public class ChartJsCallbackHandler
        {
            private ChartJs3 _ChartJs;

            public ChartJsCallbackHandler(ChartJs3 instance)
            {
                _ChartJs = instance;
            }

            [JSInvokable]
            public void Initialize()
            {
                _ChartJs.OnChartJsCreated.InvokeAsync(_ChartJs);
            }
        }

        /// <summary>
        /// Reference to chart JS Canvas
        /// </summary>
        ElementReference _ChartJSCanvas;

        /// <summary>
        /// Reference to Chart JS Instance
        /// </summary>
        IJSObjectReference _ChartJSInstance;

        /// <summary>
        /// Object Model for initializing chartjs
        /// </summary>
        [Parameter]
        public JObject ChartJsInitializer { get; set; }

        /// <summary>
        /// Height
        /// </summary>
        [Parameter]
        public double? Height { get; set; }

        /// <summary>
        /// Whether to use local Chart.js 
        /// </summary>
        [Parameter]
        public bool UseLocalChartJs { get; set; }

        [Inject]
        public IJSRuntime JSRuntime { get; set; }

        /// <summary>
        /// Called when ChartJS instance was created
        /// </summary>
        [Parameter]
        public EventCallback<ChartJs3> OnChartJsCreated { get; set; }

        private DotNetObjectReference<ChartJsCallbackHandler> _Handler;

        protected override async Task OnAfterRenderAsync(bool firstRender)
        {
            if (firstRender == false)
            {
                return;
            }

            var module = await this.JSRuntime.InvokeAsync<IJSObjectReference>("import", "./_content/NC.Blazor.ChartJs3/ChartJs3.razor.js");

            _Handler = DotNetObjectReference.Create(new ChartJsCallbackHandler(this));
            _ChartJSInstance = await module.InvokeAsync<IJSObjectReference>("ChartJS3Blazor", _ChartJSCanvas, this.ChartJsInitializer.ToString(), false, _Handler);

        }

        private void ThrowIfChartNotInitialized()
        {
            if (_ChartJSInstance == null)
            {
                throw new InvalidOperationException("Chart was not initialized");
            }
        }

        /// <summary>
        /// Changes the data in the chart
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <param name="labels"></param>
        /// <returns></returns>
        public ValueTask ChangeData(int dataSetIndex, double[] data, string[] labels = null)
        {
            this.ThrowIfChartNotInitialized();

            return _ChartJSInstance.InvokeVoidAsync("changeData", dataSetIndex, data, labels);

        }

        /// <summary>
        /// Changes the data in the chart using tuple format
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <param name="labels"></param>
        /// <returns></returns>
        public ValueTask ChangeData(int dataSetIndex, Data[] data)
        {
            this.ThrowIfChartNotInitialized();

            return _ChartJSInstance.InvokeVoidAsync("changeDataRaw", dataSetIndex, data);

        }

        /// <summary>
        /// Pushes data into the chart's data set
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <param name="labels"></param>
        /// <returns></returns>
        public ValueTask PushData(int dataSetIndex, double value, string label = null)
        {
            this.ThrowIfChartNotInitialized();

            return _ChartJSInstance.InvokeVoidAsync("pushData", dataSetIndex, value, label);
        }

        /// <summary>
        /// Pushes data into the chart's data set using tuple format
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <returns></returns>
        public ValueTask PushData(int dataSetIndex, Data data)
        {
            this.ThrowIfChartNotInitialized();

            return _ChartJSInstance.InvokeVoidAsync("pushData", dataSetIndex, data);
        }

        /// <summary>
        /// Pushes data into the chart's data set
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="values"></param>
        /// <returns></returns>
        public ValueTask PushData(int dataSetIndex, double[] values)
        {
            this.ThrowIfChartNotInitialized();

            return _ChartJSInstance.InvokeVoidAsync("pushMultipleData", dataSetIndex, values);
        }

        /// <summary>
        /// Pushes data into the chart's data set
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="values"></param>
        /// <returns></returns>
        public ValueTask PushData(int dataSetIndex, Data[] values)
        {
            this.ThrowIfChartNotInitialized();

            return _ChartJSInstance.InvokeVoidAsync("pushMultipleData", dataSetIndex, values);
        }

        /// <summary>
        /// Pushes data into the chart's data set and remove the first item. Creating "Time Window" effect.
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <returns></returns>
        public ValueTask PushDataAndShift(int dataSetIndex, double data)
        {
            this.ThrowIfChartNotInitialized();

            return _ChartJSInstance.InvokeVoidAsync("pushDataAndShift", dataSetIndex, data);
        }

        /// <summary>
        /// Pushes data into the chart's data set and remove the first item. Creating "Time Window" effect.
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <param name="labels"></param>
        /// <returns></returns>
        public ValueTask PushDataAndShift(int dataSetIndex, Data data)
        {
            this.ThrowIfChartNotInitialized();

            return _ChartJSInstance.InvokeVoidAsync("pushDataAndShift", dataSetIndex, data);
        }

        /// <summary>
        /// Remove First Item from the data set by calling Array.shift() on the data
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <param name="labels"></param>
        /// <returns></returns>
        public async ValueTask<JToken> ShiftData(int dataSetIndex)
        {
            this.ThrowIfChartNotInitialized();

            var json = await _ChartJSInstance.InvokeAsync<string>("shiftData", dataSetIndex);
            return JToken.Parse(json);
        }

        /// <summary>
        /// Remove Data by calling Array.splice
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <param name="labels"></param>
        /// <returns></returns>
        public async ValueTask<JToken> RemoveData(int dataSetIndex, int startingIndex, int howMany)
        {
            this.ThrowIfChartNotInitialized();

            var json = await _ChartJSInstance.InvokeAsync<string>("removeData", dataSetIndex, startingIndex, howMany);
            return JToken.Parse(json);
        }

        /// <summary>
        /// Gets current option of the chart
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <param name="labels"></param>
        /// <returns></returns>
        public async ValueTask<JObject> GetOptions()
        {
            this.ThrowIfChartNotInitialized();

            var json = await _ChartJSInstance.InvokeAsync<string>("getOptions");
            return JObject.Parse(json);
        }

        /// <summary>
        /// Update chart options
        /// </summary>
        /// <param name="dataSetIndex"></param>
        /// <param name="data"></param>
        /// <param name="labels"></param>
        /// <returns></returns>
        public ValueTask SetOptions(JObject newOption)
        {
            this.ThrowIfChartNotInitialized();

            return _ChartJSInstance.InvokeVoidAsync("changeOptions", newOption.ToString());
        }

        public async ValueTask DisposeAsync()
        {
            GC.SuppressFinalize(this);

            if (_ChartJSInstance != null)
            {
                await _ChartJSInstance.InvokeVoidAsync("dispose");
                await _ChartJSInstance.DisposeAsync();
            }

            _Handler?.Dispose();
        }
    }
}

File: wwwroot/ChartJs3.razor.js
export function ChartJS3Blazor(canvasReference, initializeObject, useLocal, callbackHandler ) {

    var $me = {};

    function wait(lib, name, callback) {

        if (lib != null) {

            var key = "lib-" + name;
            if (document.getElementsByClassName(key).length == 0) {

                var newScript = document.createElement("script");
                newScript.src = lib;
                newScript.className = key;
                document.getElementsByTagName("head")[0].appendChild(newScript);
            }
        }

        if (window[name]) {
            callback(window[name]);
            return;
        }

        window.setTimeout(function () {

            wait(null, name, callback);
        }, 100);
    }

    var initialize = function (ChartObj) {

        console.log("ChartJS Initializing, config:")
        var config = JSON.parse(initializeObject);

        console.log(config);

        $me.chart = new ChartObj(canvasReference, config);
        console.log("ChartJS Initialized")

        var preCondition = function (datasetIndex) {

            if ($me.chart == null) {
                console.log("Chart was destroyed or not initialized properly");
                return false;
            }

            if (datasetIndex != null) {

                if ($me.chart.data.datasets[datasetIndex] == null) {
                    console.log(datasetIndex + " is not a valid data set index, it will be automatically created. check your initialization again");
                    $me.chart.data.datasets[datasetIndex] = {
                        data: []
                    };
                }

                if ($me.chart.data.datasets[datasetIndex].data == null) {

                    $me.chart.data.datasets[datasetIndex].data = []
                }
            }

            return true;
        };

        $me.dispose = function () {

            if ($me.chart != null) {
                $me.chart.destroy();
                $me.chart = null;
            }
        };

        $me.rebuild = function (newConfig) {

            $me.dispose();
            $me.chart = new ChartObj(canvasReference, newConfig);
        };

        $me.changeData = function (datasetIndex, data, labels) {

            if (!preCondition(datasetIndex))
                return;

            if (data != null) {

                $me.chart.data.datasets[datasetIndex].data = data;
            }

            if (labels != null) {

                $me.chart.data.labels = labels;
            }

            $me.chart.update();
        };

        $me.changeAllData = function (dataArray, labels) {

            if (!preCondition())
                return;

            if (Array.isArray(data)) {

                dataArray.forEach(function (item) {
                    $me.chart.data.datasets[datasetIndex].data = item;
                });
            }

            if (Array.isArray(labels) != null) {

                $me.chart.data.labels = labels;
            }

            $me.chart.update();
        };

        $me.changeDataRaw = function (datasetIndex, data) {

            if (!preCondition(datasetIndex))
                return;

            if (Array.isArray(labels) != null) {

                $me.chart.data.datasets[datasetIndex].data = data;
            }
            $me.chart.update();
        };

        $me.pushData = function (datasetIndex, dataItem, label) {

            if (!preCondition(datasetIndex))
                return;
            
            $me.chart.data.datasets[datasetIndex].data.push(dataItem);

            if (label != null && $me.chart.data.labels == null) {

                $me.chart.data.labels = [];
            }

            $me.chart.data.labels.push(label);
            $me.chart.update();
        }

        $me.pushDataAndShift = function (datasetIndex, dataItem) {

            if (!preCondition(datasetIndex))
                return;

            $me.chart.data.datasets[datasetIndex].data.push(dataItem);
            $me.chart.data.datasets[datasetIndex].data.splice(0, 1);;
            $me.chart.update();
        }

        $me.pushMultipleData = function (datasetIndex, dataArray) {

            if (!preCondition(datasetIndex))
                return;

            dataArray.forEach(function (item) {
                $me.chart.data.datasets[datasetIndex].data.push(item);
            });

            $me.chart.update();
        }

        $me.shiftData = function (datasetIndex, index) {

            if (!preCondition(datasetIndex))
                return;

            $me.chart.data.datasets[datasetIndex].data.shift();
            $me.chart.update();
        }

        $me.removeData = function (datasetIndex, startingIndex, howMany) {

            if (!preCondition(datasetIndex))
                return;

            $me.chart.data.datasets[datasetIndex].data.splice( startingIndex, howMany );

            $me.chart.update();
        }

        $me.getOptions = function () {

            if ($me.chart == null) {
                console.log("Chart was destroyed or not initialized properly");
                return "{}";
            }

            return JSON.stringify($me.chart.options);
        };

        $me.changeOptions = function (newOptions) {

            if ($me.chart == null) {
                console.log("Chart was destroyed or not initialized properly");
                return;
            }

            $me.chart.options = JSON.parse(newOptions);
            $me.chart.update();
        };

        callbackHandler.invokeMethodAsync('Initialize');
    };

    if (useLocal) {

        wait("./_content/NC.Blazor.ChartJs3/chart.js", "Chart", initialize);
    }
    else {

        wait("https://cdn.jsdelivr.net/npm/chart.js", "Chart", initialize);
    }


    return $me;
};